data_model/src/asset.rs

```diff
@@ -413,37 +412,47 @@ impl FromStr for DefinitionId {
     FfiType,
     IntoSchema,
 )]
-#[display(fmt = "{}#{}", "self.definition_id.name", "self.account_id")]
 pub struct Id {
     /// Entity Identification.
     pub definition_id: <AssetDefinition as Identifiable>::Id,
     /// Account Identification.
     pub account_id: <Account as Identifiable>::Id,
 }
 
+impl fmt::Display for Id {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.definition_id.domain_id == self.account_id.domain_id {
+            write!(f, "{}##{}", self.definition_id.name, self.account_id)
+        } else {
+            write!(f, "{}#{}", self.definition_id, self.account_id)
+        }
+    }
+}
+
 /// Asset Identification is represented by `name#account@domain` string.
 impl FromStr for Id {
     type Err = ParseError;
 
     fn from_str(string: &str) -> Result<Self, Self::Err> {
-        let mut split = string.split('#');
-        match (split.next(), split.next(), split.next()) {
-            (Some(""), _, _) => Err(ParseError {
-                reason: "Asset ID cannot be empty",
-            }),
-            (Some(name), Some(account_id), None) if !account_id.is_empty() => {
-                let name = name.parse()?;
-                let account_id = <Account as Identifiable>::Id::from_str(account_id)?;
-                let definition_id =
-                    <AssetDefinition as Identifiable>::Id::new(name, account_id.domain_id.clone());
-                Ok(Self {
-                    definition_id,
-                    account_id,
-                })
-            }
-            _ => Err(ParseError {
-                reason: "Asset ID should have format `asset#account@domain`",
-            }),
+        if let Some((asset_definition_candidate, account_id_candidate)) = string.rsplit_once('#') {
+            let account_id: <Account as Identifiable>::Id = account_id_candidate.parse().map_err(|_err| ParseError{reason: "Failed to parse the `account_id` part of the `asset_id`. Please ensure that it has the form `account@domain`"})?;
```

Comment:
It seems errors in other `FromStr` impls don't assign responsibility to the user, just state what the string should've looked like.

Reply:
IMO that makes sense, since the person seeing this error isn't necessarily the one who supplied the string.